File: results/RQ1/05_baseline_qualification/units/p07_calculator/java/attempts/0001_seed_1570566783/p07_calculator_java/p07_calculator_java/src/main/java/Tokenizer.java
/**
 * Tokenizer for arithmetic expressions.
 * Converts a string input into a list of tokens.
 * Skips whitespace and recognizes numbers, operators, and parentheses.
 */
import java.util.ArrayList;
import java.util.List;

public class Tokenizer {
    /** The input string being tokenized */
    private final String input;
    
    /** Current position in the input string */
    private int position;
    
    /** Length of the input string */
    private final int length;

    /**
     * Constructs a new Tokenizer with the given input string.
     *
     * @param input the arithmetic expression to tokenize
     */
    public Tokenizer(String input) {
        this.input = input;
        this.position = 0;
        this.length = input.length();
    }

    /**
     * Tokenizes the input string into a list of tokens.
     *
     * @return a list of tokens representing the input expression
     * @throws TokenizationException if the input contains invalid characters or malformed numbers
     */
    public List<Token> tokenize() throws TokenizationException {
        List<Token> tokens = new ArrayList<>();
        
        while (position < length) {
            char ch = input.charAt(position);
            
            // Skip whitespace
            if (Character.isWhitespace(ch)) {
                position++;
                continue;
            }
            
            // Handle numbers (including decimals)
            if (Character.isDigit(ch) || ch == '.') {
                tokens.add(readNumber());
            } 
            // Handle operators and parentheses
            else if (ch == '+') {
                tokens.add(new Token(TokenType.PLUS, "+", null));
                position++;
            } else if (ch == '-') {
                tokens.add(new Token(TokenType.MINUS, "-", null));
                position++;
            } else if (ch == '*') {
                tokens.add(new Token(TokenType.STAR, "*", null));
                position++;
            } else if (ch == '/') {
                tokens.add(new Token(TokenType.SLASH, "/", null));
                position++;
            } else if (ch == '(') {
                tokens.add(new Token(TokenType.LPAREN, "(", null));
                position++;
            } else if (ch == ')') {
                tokens.add(new Token(TokenType.RPAREN, ")", null));
                position++;
            } 
            // Invalid character
            else {
                throw new TokenizationException("Invalid character '" + ch + "' at position " + position);
            }
        }
        
        // Add EOF token
        tokens.add(new Token(TokenType.EOF, "", null));
        return tokens;
    }

    /**
     * Reads a number token starting at the current position.
     *
     * @return a Token representing the number
     * @throws TokenizationException if the number is malformed
     */
    private Token readNumber() throws TokenizationException {
        int start = position;
        
        // Read digits and decimal point
        while (position < length && (Character.isDigit(input.charAt(position)) || input.charAt(position) == '.')) {
            position++;
        }
        
        // Extract the number string
        String numberStr = input.substring(start, position);
        
        // Validate and convert to Double
        try {
            Double numericValue = Double.parseDouble(numberStr);
            return new Token(TokenType.NUMBER, numberStr, numericValue);
        } catch (NumberFormatException e) {
            throw new TokenizationException("Malformed number '" + numberStr + "' at position " + start);
        }
    }
}

File: results/RQ1/05_baseline_qualification/units/p07_calculator/python/attempts/0001_seed_2130722843/p07_calculator_python/p07_calculator_python/calculator/tokenizer.py
"""Tokenizer for arithmetic expressions.

This module provides the Token classes and the Tokenizer class
for converting arithmetic expressions into tokens.
"""

from typing import List, Union
from .errors import TokenizationError


class Token:
    """Marker base class for all token types."""


class NumberToken(Token):
    """Token representing a numeric value."""

    def __init__(self, value: float) -> None:
        """Initialize a NumberToken with a numeric value.

        Args:
            value: The numeric value represented by this token.
        """
        self.value = value


class OpToken(Token):
    """Token representing an operator."""

    def __init__(self, op: str) -> None:
        """Initialize an OpToken with an operator.

        Args:
            op: The operator symbol (+, -, *, /).

        Raises:
            ValueError: If the operator is not one of +, -, *, /.
        """
        if op not in "+-*/":
            raise ValueError(f"Invalid operator: {op}")
        self.op = op


class LParenToken(Token):
    """Token representing a left parenthesis."""

    def __init__(self) -> None:
        """Initialize an LParenToken."""
        pass


class RParenToken(Token):
    """Token representing a right parenthesis."""

    def __init__(self) -> None:
        """Initialize an RParenToken."""
        pass


class Tokenizer:
    """Converts arithmetic expressions into lists of tokens."""

    def tokenize(self, expression: str) -> List[Token]:
        """Convert an expression string into a list of tokens.

        Args:
            expression: A string containing a valid arithmetic expression.

        Returns:
            A list of tokens representing the expression.

        Raises:
            TokenizationError: If the expression contains invalid characters
                or malformed numbers.
        """
        if not expression:
            raise TokenizationError("Empty expression")

        tokens: List[Token] = []
        i = 0

        while i < len(expression):
            char = expression[i]

            # Skip whitespace
            if char.isspace():
                i += 1
                continue

            # Handle numbers (including decimals)
            if char.isdigit() or char == '.':
                start = i
                while i < len(expression) and (
                    expression[i].isdigit() or expression[i] == '.'
                ):
                    i += 1
                num_str = expression[start:i]
                try:
                    # Try integer first, then float
                    if '.' in num_str:
                        value = float(num_str)
                    else:
                        value = int(num_str)
                    tokens.append(NumberToken(value))
                except ValueError:
                    raise TokenizationError(
                        f"Malformed number: {num_str}"
                    )
                continue

            # Handle unary minus or plus
            if char in "+-" and (
                i == 0 or
                expression[i-1] in "(+-*/"
            ):
                # Look ahead to see if next character is a digit
                j = i + 1
                while j < len(expression) and expression[j].isspace():
                    j += 1
                if j < len(expression) and expression[j].isdigit():
                    # Found unary operator followed by digit, treat as negative
                    # We'll process the number part in the next iteration
                    # Add the operator as a token
                    tokens.append(OpToken(char))
                    i += 1
                    continue
                elif j < len(expression) and expression[j] == '.':
                    # Check if it's a decimal number starting with unary minus/plus
                    k = j + 1
                    while k < len(expression) and expression[k].isdigit():
                        k += 1
                    if k < len(expression) and expression[k] == '.':
                        # Invalid - more than one decimal point
                        raise TokenizationError(
                            f"Malformed number: {expression[i:j+1]}"
                        )
                    elif k == len(expression) or not expression[k].isdigit():
                        # This case handles unary plus/minus followed by decimal
                        # but we don't want to consume the decimal yet
                        # Let it be handled by the number detection logic
                        pass
                    else:
                        # Invalid decimal format
                        raise TokenizationError(
                            f"Malformed number: {expression[i:j+1]}"
                        )

            # Handle operators and parentheses
            if char in "+-*/()":
                if char == '(':
                    tokens.append(LParenToken())
                elif char == ')':
                    tokens.append(RParenToken())
                else:
                    tokens.append(OpToken(char))
                i += 1
                continue

            # Everything else is invalid
            raise TokenizationError(
                f"Unexpected character: '{char}'"
            )

        return tokens